subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/IncrementalTaskAction.java

```diff
@@ -54,39 +51,29 @@ public void releaseContext() {
     }
 
     protected void doExecute(final Task task, String methodName) {
-        IncrementalTaskInputs incrementalInputs = context.getExecutionStateChanges()
-            .map(new Function<ExecutionStateChanges, StatefulIncrementalTaskInputs>() {
-                @Override
-                public StatefulIncrementalTaskInputs apply(ExecutionStateChanges changes) {
-                    return changes.getInputFilesChanges().map(new Function<Iterable<Change>, StatefulIncrementalTaskInputs>() {
-                        @Override
-                        public StatefulIncrementalTaskInputs apply(Iterable<Change> changes) {
-                            return createIncrementalInputs(changes);
-                        }
-                    }).orElseGet(new Supplier<StatefulIncrementalTaskInputs>() {
-                        @Override
-                        public StatefulIncrementalTaskInputs get() {
-                            return createRebuildInputs(task);
-                        }
-                    });
-                }
-            }).orElseGet(new Supplier<StatefulIncrementalTaskInputs>() {
-                @Override
-                public StatefulIncrementalTaskInputs get() {
-                    return createRebuildInputs(task);
-                }
-            });
+        @SuppressWarnings("OptionalGetWithoutIsPresent")
+        ExecutionStateChanges changes = context.getExecutionStateChanges().get();
```

Comment:
Can we avoid doing this?

Reply:
Avoid what exactly? Calling `get` on the `Optional`?

I can add a condition and throw an exception when the optional is empty. I think it should be an error trying to execute an incremental task action when the task wants to ignore changes.

Reply:
I agree that it should be an error, but I think it should be a compile error. We shouldn't have to deal with the case of no changes here. For now I'd change `TaskExecutionContext.getExecutionStateChanges()` to return a non-optional result.